### PR TITLE
Calendar Events

### PR DESCRIPTION
**Pull request by Nataliazeus94**
Contains the implementation of the Calendar events

### CODE REVIEW

**Comment by Nataliazeus94** [comment]
Calendar Event implemention

File: modules/routes/calendar/calendar_event_route.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+@app.route("/calendar", methods=["POST"])
+@protected_route(roles=[AccountType.Admin, AccountType.Lecturer])
+def create_calendar_event():
+    body = CalendarEventSchema().load(request.get_json(force=True))
+
+    db_cursor = db.cursor()
+    try:
+        db_cursor.execute(
+            "INSERT INTO CalendarEvent (course_id, date, event_name, event_no) VALUES (%s, %s, %s, %s)",
```
Comment (bombies) [inline modules/routes/calendar/calendar_event_route.py]: How are you inserting the event_no into the record? That's an auto-generated primary key. Omit that from insertion and remove it from the payload schema.

File: modules/routes/calendar/calendar_event_route.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+from app import app
+from modules.models.account import AccountType
+from modules.routes.calendar.calendar_event_schema import (
+    CalendarEventSchema,
+
+)
+from modules.utils.db import db
+from modules.utils.route_utils import authenticate, fetch_session, protected_route
+
+@app.route("/calendar", methods=["POST"])
```
Comment (bombies) [inline modules/routes/calendar/calendar_event_route.py]: The course_id should be in the route. Remove it from payload schema.

```suggestion
@app.route("/calendar/<str:course_id>", methods=["POST"])
```

File: modules/routes/calendar/calendar_event_route.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+@protected_route(roles=[AccountType.Admin, AccountType.Lecturer])
+def create_calendar_event():
```
Comment (bombies) [inline modules/routes/calendar/calendar_event_route.py]:
```suggestion
def create_calendar_event(course_id: str):
```

File: modules/routes/calendar/calendar_event_route.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    db_cursor = db.cursor()
+    db_cursor.execute("SELECT * FROM CalendarEvent WHERE course_id = %s", (course_id,))
+    events = db_cursor.fetchall()
+    if not events:
+        return jsonify({"message": "No calendar events found for this course"}), 404
+    # Convert the events data into JSON format and return
+    return jsonify(events), 200
+    # Return data as per requirements
+
+@app.route("/calendar/events/student/<int:student_id>/date/<date:date>", methods=["GET"])
```
Comment (bombies) [inline modules/routes/calendar/calendar_event_route.py]: This route is too long and it limits what the endpoint could do. Instead, create a schema to allow a startDate and an endDate. The endpoint will handle the results according to said schema.

```suggestion
@app.route("/calendar/events/student/<int:student_id>", methods=["GET"])
```

File: modules/routes/calendar/calendar_event_route.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+def get_student_calendar_events(student_id, date):
```
Comment (bombies) [inline modules/routes/calendar/calendar_event_route.py]:
```suggestion
def get_student_calendar_events(student_id):
```

File: modules/routes/calendar/calendar_event_route.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    db_cursor = db.cursor()
+    db_cursor.execute("SELECT * FROM CalendarEvent WHERE date = %s AND course_id IN (SELECT course_id FROM Enrollment WHERE student_id = %s)", (date, student_id))
```
Comment (bombies) [inline modules/routes/calendar/calendar_event_route.py]: The payload schema should allow for both an optional start date and an optional end date. Update query to select all events if there is no start/end date provided. If only one field is provided, filter accordingly and if both are provided, select within the range.

File: modules/routes/calendar/calendar_event_schema.py [hunk 7]
```diff
@@ -0,0 +1,7 @@
+from marshmallow import Schema, fields
+
+class CalendarEventSchema(Schema):
+    course_id = fields.String(required=True)
+    date = fields.Date(required=True)
+    event_name = fields.String(required=True)
+    event_no = fields.Int(required=True)
```
Comment (bombies) [inline modules/routes/calendar/calendar_event_schema.py]:
```suggestion
```

**Review by bombies — CHANGES_REQUESTED** [review]
See changes and re-implement ASAP.

File: modules/routes/calendar/calendar_event_route.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+@app.route("/calendar/<str:course_id>", methods=["POST"])
+@protected_route(roles=[AccountType.Admin, AccountType.Lecturer])
+def create_calendar_event(course_id: str):
+    body = CalendarEventSchema().load(request.get_json(force=True))
+
+    db_cursor = db.cursor()
+    try:
+        db_cursor.execute(
+            "INSERT INTO CalendarEvent (course_id, date, event_name, event_no) VALUES (%s, %s, %s, %s)",
```
Comment (bombies) [inline modules/routes/calendar/calendar_event_route.py]:
```suggestion
            "INSERT INTO CalendarEvent (course_id, date, event_name) VALUES (%s, %s, %s)",
```

File: modules/routes/calendar/calendar_event_route.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+            (body["course_id"], body["date"], body["event_name"]),
+        )
+        db.commit()
+        return jsonify({"message": "Calendar event created successfully"}), 201
+    except Exception as e:
+        db.rollback()
+        return jsonify({"message": f"Failed to create calendar event: {str(e)}"}), 500
+
+@app.route("/calendar/events/course/<string:course_id>", methods=["GET"])
+def get_course_calendar_events(course_id):
```
Comment (bombies) [inline modules/routes/calendar/calendar_event_route.py]: This should be a protected route. I'm almost certain I told you all the endpoints in this route should be protected...

```suggestion
@protected_route()
def get_course_calendar_events(course_id):
```

There should also be checks to see if a student is enrolled in the course or if a lecturer teaches the course.

File: modules/routes/calendar/calendar_event_route.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    db_cursor.execute("SELECT * FROM CalendarEvent WHERE course_id = %s", (course_id,))
+    events = db_cursor.fetchall()
+    if not events:
+        return jsonify({"message": "No calendar events found for this course"}), 404
+    # Convert the events data into JSON format and return
+    return jsonify(events), 200
+    # Return data as per requirements
+
+@app.route("/calendar/events/student/<int:student_id>", methods=["GET"])
+def get_student_calendar_events(student_id):
```
Comment (bombies) [inline modules/routes/calendar/calendar_event_route.py]:
```suggestion
@protected([AccountType.Student, AccountType.Admin])
def get_student_calendar_events(student_id):
```

There should also be checks to see if the student ID from the account ID in the session matches the ID passed to the function if the session is a student account.

**Review by bombies — CHANGES_REQUESTED** [review]

**Review by bombies — APPROVED** [review]

**Pull request merged**